[ADD][sale_block_cancellation] Module to block cancellation of sales with validation added.

luistorresm opened the pull request:
Added a new button in Sale Orders to only allow cancel a SO when all the
pickings related with this are on `ready to transfer` or in
`waiting for another move` state, or even with pickings on different states
if the moved product quantities are correctly returned.

The original button to cancel is added to a new group `Force Sale Cancel`
to can allow cancel a Sale Order out of these cases.
- [x] Added module to block cancellation of sale orders.
- [x] Added tests

nhomar commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,53 @@
+            for pick in sale.picking_ids:
+                if pick.state == 'done':
+                    for move in pick.move_lines:
+                        prod_id = move.product_id.id
```
Por favor por este algoritmo (el de las ubicaciones.) en un micro metodo externo para bajar la complejidad ciclomática.

nhomar replied:
desde la 28 hasta la 42 puede ir en un metodo externo.

nhomar commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,53 @@
+        allow = True
+        for sale in self:
+            dict_prod = {}
+            for pick in sale.picking_ids:
```
Una vez tengas un método externo usa una list comprehension aqui.

nhomar commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,53 @@
+                    "are in 'Ready to transfer' or 'Waiting another move'."
+                    "\n- If the moved product quantities in pickings are "
+                    "correctly returned."))
+        return self.action_cancel()
```
me gusta más si sobreescribes y preprocesas el cancel original ¿Por qué un botón por separado aquí y no un super preprocesado.?

nhomar commented:
No veo test unitarios.

sabrinagnu commented on sale_block_cancellation/security/res_groups.xml:
```diff
@@ -0,0 +1,10 @@
+    <data noupdate="0">
+
+        <record id="group_cancel_sale_blocked" model="res.groups">
+            <field name="name">Cancel Sales Blocked</field>
```
@luistorresm cambiar "Cancel Sales Blocked" >> Show default sale order cancel button

sabrinagnu replied:
@luistorresm y en español "Mostrar botón cancelar por defecto en SO"

luistorresm commented:
Done @dsabrinarg

luistorresm commented:
@JulioSerna 

May you give me +1?

JulioSerna commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,108 @@
+        string='Picking associated to this sale', search=_search_pickings)
+
+    @api.multi
+    def check_prod_by_locations(self):
```
@luistorresm 

What happen is the warehouse is in three steps to delivery and the customer make a return and return is of customer to stock directly , this work as well?

luistorresm commented:
@JulioSerna 

The sale order allow cancel, because all the moves are correctly returned in that case.

JulioSerna commented on sale_block_cancellation/tests/test_sale_block_cancellation.py:
```diff
@@ -0,0 +1,83 @@
+        self.proc_obj.create({}).procure_calculation()
+        for picking in sale.picking_ids:
+            if picking.state == 'confirmed':
+                picking.force_assign()
```
@luistorresm 

You must avoid to force assing,

https://github.com/Vauxoo/lodigroup/pull/451

luistorresm replied:
Done @JulioSerna

sabrinagnu commented:
@luistorresm "The sale order allow cancel, because all the moves are correctly returned in that case.", Which test you did to affirm this?, I believe this is not true and the system is not allowing the cancellation on the case that Julio pointed.

luistorresm commented:
@JulioSerna, 

@dsabrinarg have reason, if I try cancel a SO when this are Done, I get the next message:
`You cannot cancel a sale order line which is linked to a stock move already done.`

JulioSerna commented:
Pending to review @dsabrinarg  with @nhomar

josemoralesp commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,108 @@
+            for loc in dict_prod.get(prod):
+                if dict_prod.get(prod).get(loc) != 0:
+                    return False
+        return True
```
Hello @luistorresm 

Please can you make a refactor for this method.

Please Consider the use of sql, because if we have orders with 300 lines(like a normal day in apex) we will have more than 900 lines considering the picking returned.

This method is called when we install this module because the compute field is store=True.

This module may take many hours installing for this reason.

Please let me know any doubt with this message

Best Regards.

luistorresm replied:
Can check again @josemoralesp .

I was change to query.

josemoralesp replied:
Hello @luistorresm 

The main reason for the sql is avoid the excessive use of sentence 'for' in your method.

I was Checking your method and you are using the move to know if the products were correctly returned but with this you need to do something else(a lot of 'for' sentences).

I want to suggest you the use of the quant instead the move, because if the location of the quants used for your moves is different to the origin of the move(It may be stock) you can say that you have products without return.

E.G.

``` python
self._cr.execute('''
SELECT product_id, 
       sum(qty),
       location_id 
FROM stock_quant 
WHERE id IN (SELECT quant_id 
             FROM stock_quant_move_rel 
             WHERE move_id IN (SELECT id 
                               FROM stock_move 
                               WHERE picking_id IN %s)) AND
       location_id != %s 
GROUP BY product_id, location_id
''', (tuple(pick_ids.ids), main_loc_id))
```

Where main_loc_id is the location where products initiated the moves.

If this sql sentences returns something this means that you have products without returns, no matter if they return the moves directly from customer to stock.

Do you understand my point?

josemoralesp commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,111 @@
+        allowed, or even whit pickings on different states, if the moved
+        product quantities are correctly returned.'''
+        for sale in self:
+            sale.allow_cancel = False
```
By default the boolean fields are False, you can exclude this to avoid call the write method twice for each order

luistorresm replied:
Done @josemoralesp

luistorresm committed:
[ADD][sale_block_cancellation] Module to block cancellation of sales with validation added

Added a new button in Sale Orders to only allow cancel a SO when all the
pickings related with this are on ready to transfer or in
waiting for another move state, or even with pickings on different
states if the moved product quantities are correctly returned.

The original button to cancel is added to a new group Force Sale Cancel
to can allow cancel a Sale Order out of these cases.

josemoralesp commented on sale_block_cancellation/models/sale.py:
```diff
@@ -0,0 +1,109 @@
+        for prod in prods:
+            if sum(prod[1]) != sum(prod[2]):
+                return False
+        return True
```
Perfect

luistorresm replied:
@josemoralesp

Can you give me an +1?

Regarda

luistorresm committed:
[REF][sale_block_cancellation]Solved pylint

josemoralesp commented:
:+1:
